core/modules/wcontrol/TransmitMgr.cc

```diff
@@ -57,49 +57,23 @@ int TransmitMgr::getTransmitCount(qmeta::CzarId czarId) const {
 }
 
 
-int TransmitMgr::getAlreadyTransCount(qmeta::CzarId czarId) const {
-    unique_lock<mutex> uLock(_mtx);
-    auto const& iter = _czarTransmitMap.find(czarId);
-    if (iter != _czarTransmitMap.end()) {
-        return iter->second._alreadyTransCount;
-    }
-    return 0;
-}
-
-
 void TransmitMgr::_take(bool interactive, bool alreadyTransmitting, qmeta::CzarId czarId) {
     unique_lock<mutex> uLock(_mtx);
     LOGS(_log, LOG_LVL_DEBUG, "TransmitMgr take locking " << dump());
     TransmitInfo& info = _czarTransmitMap[czarId];
+
     ++(info._takeCalls);
     ++(info._totalCount);
-    // Check if the caller needs to wait.
-    // This is trying to prioritize transmits that are 'alreadyTransmitting' so
-    // they finish and stop using system resources on the czar (with czarId).
-    // '_maxAlreadyTran' should be significantly smaller than '_maxTransmits', which
-    // causes all new transmits to have to wait until some of the already transmitting
-    // items have cleared.
-    // '_maxTransmits' may be 50 while '_maxAlreadyTran' may be 10.
-    // Interactive transmits never need to wait.
-    if (not interactive
-        || info._transmitCount >= _maxTransmits
-        || info._alreadyTransCount >= _maxAlreadyTran) {
-        // If not alreadyTransmitting, it needs to wait until the number of already transmitting
-        // jobs drops below _maxAlreadyTran before it can start transmitting.
-        if (alreadyTransmitting) {
-            ++(info._alreadyTransCount);
-            LOGS(_log, LOG_LVL_DEBUG, "czar=" << czarId
-                 << " ++_alreadyTransCount=" << info._alreadyTransCount);
+    if (not interactive) {
+        // Check if the caller needs to wait.
+        // Interactive transmits never need to wait.
+        if (info._transmitCount >= _maxTransmits) {
             _tCv.wait(uLock, [this, &info](){ return info._transmitCount < _maxTransmits; });
-        } else {
-            _tCv.wait(uLock, [this, &info](){
-                return (info._transmitCount < _maxTransmits)
-                        && (info._alreadyTransCount < _maxAlreadyTran);
-            });
         }
+        ++(info._transmitCount);
```

Comment:
In the previous version of the code the interactive transmits were also being counted. The new version seems to exclude them. Was this done on purpose?

Reply:
Yes, this was the source of the lockup problem. There was an effort to prioritize large results, but either the logic here was faulty and releases were not happening correctly, and/or the interactions between TransmitMgr, SqlConnectionMgr, and double buffering could cause the system could lock up.

After simplifying the logic here, I've been unable to reproduce the lockup. Prioritizing large results was probably of limited value.